File: src/problemas/grafos/1082/1082.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;

public class Main {
	public static void DFS(ArrayList<Integer> visitado, ArrayList<ArrayList<Integer>> grafo, int componente, int u){
		visitado.set(u, componente);

		for(int i = 0; i < grafo.get(u).size(); ++i){
			int v = grafo.get(u).get(i);

			if(visitado.get(v) == 0){
				DFS(visitado, grafo, componente, v);
			}
		}
	}
	
    public static void main(String[] args) throws IOException {
        InputStreamReader ir = new InputStreamReader(System.in);
        BufferedReader in = new BufferedReader(ir);

        int N = Integer.parseInt(in.readLine());
		for(int k = 1; k <= N; ++k){
			String[] VE = in.readLine().trim().split(" ");
			int V = Integer.parseInt(VE[0]);
			int E = Integer.parseInt(VE[1]);

			ArrayList<Integer> visitado = new ArrayList<Integer>(); 
			ArrayList<ArrayList<Integer>> grafo = new ArrayList<ArrayList<Integer>>();
			for(int i = 0; i < V; ++i){
				visitado.add(0);
				grafo.add(new ArrayList<Integer>());
			}

			for(int i = 0; i < E; ++i){
				String[] uv = in.readLine().trim().split(" ");
				char u = uv[0].charAt(0);
				char v = uv[1].charAt(0);

				grafo.get(u - 'a').add(v - 'a');
				grafo.get(v - 'a').add(u - 'a');
			}

			System.out.printf("Case #%d:\n", k);
			int componentes = 0;
			for(int u = 0; u < V; ++u){
				if(visitado.get(u) == 0){
					componentes += 1;
					DFS(visitado, grafo, componentes, u);

					ArrayList<String> vertices = new ArrayList<String>();
					for(int i = 0; i < V; ++i){
						if(visitado.get(i) == componentes){
							vertices.add(Character.toString((char)(i + (int)'a')));
						}
					}

					Collections.sort(vertices);

					System.out.printf(String.join(",", vertices));
					System.out.println(",");
				}
			}
			System.out.printf("%d connected components\n\n", componentes);
		}
    }
}

File: src/problemas/grafos/1082/1082.cs
using System;
using System.Collections.Generic;
using System.Linq;

class URI
{
    static void DFS(ref List<int> visitado, ref List<List<int>> grafo, int componente, int u)
    {
        visitado[u] = componente;

        for (int i = 0; i < grafo[u].Count; ++i)
        {
            int v = grafo[u][i];

            if (visitado[v] == 0)
            {
                DFS(ref visitado, ref grafo, componente, v);
            }
        }
    }

    static void Main(string[] args)
    {
        int N = int.Parse(Console.ReadLine());
        for (int k = 1; k <= N; ++k)
        {
            List<int> VE = Console.ReadLine().Trim().Split(' ').Select(x => int.Parse(x)).ToList();
            int V = VE[0];
            int E = VE[1];

            List<int> visitado = new List<int>();
            List<List<int>> grafo = new List<List<int>>();
            for (int i = 0; i < V; ++i)
            {
                visitado.Add(0);
                grafo.Add(new List<int>());
            }

            for (int i = 0; i < E; ++i)
            {
                List<char> uv = Console.ReadLine().Trim().Split(' ').Select(x => char.Parse(x)).ToList();
                int u = uv[0];
                int v = uv[1];

                grafo[u - 'a'].Add(v - 'a');
                grafo[v - 'a'].Add(u - 'a');
            }

            Console.WriteLine($"Case #{k}:");
            int componentes = 0;
            for (int u = 0; u < V; ++u)
            {
                if (visitado[u] == 0)
                {
                    componentes += 1;
                    DFS(ref visitado, ref grafo, componentes, u);

                    List<char> vertices = new List<char>();
                    for (int i = 0; i < V; ++i)
                    {
                        if (visitado[i] == componentes)
                        {
                            vertices.Add((char)(i + 'a'));
                        }
                    }

                    vertices.Sort();

                    Console.WriteLine($"{string.Join(",", vertices)},");
                }
            }
            Console.WriteLine($"{componentes} connected components");
            Console.WriteLine("");
        }
    }
}